nav2_toolkit/src/pose_saver_node.cpp

```diff
@@ -0,0 +1,251 @@
+  this->declare_parameter<bool>("auto_restore_pose", true);
+
+  pose_file_path_ = this->get_parameter("pose_file_path").as_string();
+  double interval_sec = this->get_parameter("save_interval_sec").as_double();
+  bool auto_start = this->get_parameter("auto_start_saving").as_bool();
+  auto_restore_ = this->get_parameter("auto_restore_pose").as_bool();
+
+  timer_ = this->create_wall_timer(
+    std::chrono::duration<double>(interval_sec),
+    std::bind(&PoseSaverNode::timer_callback, this));
+  timer_->cancel();
+
+  sub_ = this->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    "amcl_pose", 10,
+    std::bind(&PoseSaverNode::pose_callback, this, std::placeholders::_1));
+
+  initial_pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>("initialpose", 10);
+  last_sub_count_ = 0;
+  amcl_monitor_timer_ = this->create_wall_timer(
+      2s, 
+      std::bind(&PoseSaverNode::amcl_monitor_callback, this));
+
+  start_service_ = this->create_service<std_srvs::srv::Trigger>(
+    "start_pose_saver",
+    std::bind(&PoseSaverNode::start_service_cb, this, std::placeholders::_1, std::placeholders::_2));
+
+  stop_service_ = this->create_service<std_srvs::srv::Trigger>(
+    "stop_pose_saver",
+    std::bind(&PoseSaverNode::stop_service_cb, this, std::placeholders::_1, std::placeholders::_2));
+
+  restore_service_ = this->create_service<std_srvs::srv::Trigger>(
+    "localise_at_last_known_position",
+    std::bind(&PoseSaverNode::restore_service_cb, this, std::placeholders::_1, std::placeholders::_2));
+
+  if (auto_start) {
+    saving_active_ = true;
+    timer_->reset();
+    RCLCPP_INFO(this->get_logger(), "Pose saving auto-started on launch.");
+  }
+
+  if (auto_restore_) {
+    RCLCPP_INFO(this->get_logger(), "Auto-restore enabled. Waiting for AMCL to become active...");
+  
+    if (this->get_parameter_or("use_sim_time", false)) {
+      RCLCPP_INFO(this->get_logger(), "Waiting for /clock");
+    }
+  
+    rclcpp::Time start = this->now();
+    rclcpp::Duration timeout = rclcpp::Duration::from_seconds(15.0);
+    while (initial_pose_pub_->get_subscription_count() == 0 &&
+           (this->now() - start) < timeout)
+    {
+      RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
+                           "Waiting for AMCL");
+    }
+  
+    if (initial_pose_pub_->get_subscription_count() == 0) {
+      RCLCPP_ERROR(this->get_logger(), "Timeout: AMCL did not become active.");
+    } else {
+      restore_pose_from_file_and_publish();
+    }
+  }
+  RCLCPP_INFO(this->get_logger(), "Pose Saver Node Initialized.");
+}
+
+void PoseSaverNode::pose_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  last_pose_ = std::make_shared<geometry_msgs::msg::PoseWithCovarianceStamped>(*msg);
+}
+
+void PoseSaverNode::timer_callback()
+{
+  if (!saving_active_ || !last_pose_) return;
+
+  try {
+    write_pose_to_file(pose_file_path_);
```

Comment:
Why not write to file when AMCL publishes to the pose topic? We could always throttle that to store every `N` seconds in case it is a fast publishing topic like from VSLAM, but then we could remove this timer entirely.